### PR TITLE
Add a codefixer to refactor HttpContext.Current

### PR DESCRIPTION
**Pull request by twsouthwick**
This adds an additional codefixer for HttpContext.Current that helps to
refactor its usage to instead use method injection. The codefixer will
update calling methods to pass in HttpContext.Current, which can be
called multiple times to move HttpContext.Current up the stack.

### CODE REVIEW

**Comment by twsouthwick** [comment]
The VB tests are disabled until @mjrousos PR is merged in (#534)

**Comment by twsouthwick** [comment]
Nm - #534 wasn't enabling VB for that analyzer. I've added an operation level analyzer for now for VB so that it'll unblock usage.

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs:109]: is this the same semantic model used above?

File: tests/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Test/HttpContextRefactor/VisualBasicHttpContextRefactorCodeFixProviderTests.cs
Comment (KSchlobohm) [inline tests/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Test/HttpContextRefactor/VisualBasicHttpContextRefactorCodeFixProviderTests.cs:387]: seems to be a theory

File: tests/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Test/HttpContextRefactor/CSharpHttpContextRefactorCodeFixProviderTests.cs
Comment (KSchlobohm) [inline tests/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Test/HttpContextRefactor/CSharpHttpContextRefactorCodeFixProviderTests.cs:114]: currentContext seems like a relatively common name.

is it easy to enumerate the params already on the method and determine if we should use currentContext+{instanceCount}?

File: tests/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Test/Verifiers/IAnalyzerTest.cs
Comment (KSchlobohm) [inline tests/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Test/Verifiers/IAnalyzerTest.cs:12]: you spoiled me with comments in other files that explained what the diagnostic was

File: tests/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Test/Verifiers/VisualBasicCodeFixVerifier.cs
Comment (KSchlobohm) [inline tests/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Test/Verifiers/VisualBasicCodeFixVerifier.cs:14]: seems to be a lot of test plumbing code in this PR. is there enough to make a test.common library?

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs:27]: should we just make a new base class we reuse? I think we're putting this in every codefixer

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs:62]: we're in the code fixer. is this check still necessary? seems like something the analyzer should handle. if there is no enclosingMethodOperation then we wouldn't raise the diagnostic

**Comment by KSchlobohm** [comment]
i think we would want to update the CHANGELOG.md to describe the different behavior for HttpContext.Current

**Comment by twsouthwick** [comment]
@CyrusNajmabadi I'd love your feedback on this codefixer. The goal of it is to move HttpContext.Current usage up the levels of a callstack via method injection (potentially more later), so it does some heavy lifting in the codefixer and not sure if there's better patterns I could explore there. Thanks!

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs
Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs:109]: Good catch. Passing it through...

File: tests/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Test/Verifiers/VisualBasicCodeFixVerifier.cs
Comment (twsouthwick) [reply tests/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Test/Verifiers/VisualBasicCodeFixVerifier.cs:14]: Maybe - I'm planning on seeing if we can just integrate some of it into the Microsoft.CodeAnalysis.Testing library itself

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs
Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs:62]: It's used to identify where to add parameters

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs
Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/HttpContextRefactorCodeFixProvider.cs:151]: Not sure - I assumed they'd be cached. @CyrusNajmabadi is this ok to do?

**Comment by CyrusNajmabadi** [comment]
In vacation. Will try to take a look when I can though :-)

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/CSharpHttpContextRefactorCodeFixProvider.cs
Comment (mjrousos) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/CSharpHttpContextRefactorCodeFixProvider.cs:14]: Can we just check whether the operation is `CS.IMethodBodyOperation` or `VB.IBlockOperation` in the base class so that we only have one code fix provider? I don't mind have a base class with two language-specific children if we need to, but these differences are so small it seems like it may be better to just have separate code paths in these two places in the base class.

**Review by mjrousos — APPROVED** [review]
This code fix looks for an existing property/parameter or adds a parameter with the same type as the original property had. So we'll be updating methods to accept System.Web.HttpContext parameters. Eventually, we don't want people running UA to depend on System.Web.HttpContext, though. I can think of two solutions to this:

1. Use System.Web.HttpContext but make sure this code fix runs earlier in the upgrade process (before TFM change, though we don't currently have a good way to do that). Then, after TFM change, have a different code fix that replaces all System.Web.HttpContext usage with Microsoft.AspNetCore.HttpContext.
2. Update this code fix to replace HttpContext.Current with properties/parameters of type Microsoft.AspNetCore.HttpContext. This would fit better with the current architecture of the tool since it would make sense to run at the same time as the other code fixes, but it has the drawback of not having System.Web symbols available (though we would have Microsoft.AspNetCore.* symbols).

Probably the first option is more useful when this is used as a standalone analyzer/code fix, but the second one works better in the context of UA. In fact, I'm not sure this would work in UA anyhow because applying it is likely to add more diagnostics than it fixes the first few times the code fix is applied and that might trip up the code fix upgrade step (which expects the number of diagnostics to decrease as code fixes are applied). 

Maybe the best option is two have two separate code fixes - this one which is used manually or in a yet-to-be-implemented pre-TFM-change UA step (though we'd have to be careful because we try harder not to break things at that point in UA's execution) and a separate but similar code fix that translates to Microsoft.AspNetCore.HttpContext after the TFM change.

**Comment by KSchlobohm** [comment]
Regarding @mjrousos 's comment. If this doesn't change types to Microsoft.AspNetCore.HttpContext then isn't this a feature regression?

**Comment by KSchlobohm** [comment]
We should be able to fix the type by running the new TypeUpgradeAnalyzer after this one runs. More than a few times now I've started wondering if we need to support sort order by declaring dependencies for Analyzers like we do for steps.

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs
Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs:64]: I was punting that till later so I could just get it in, but I went back and made it work with syntax level and tried to reuse what I could of the current one.

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/NameMatcher.cs
Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/NameMatcher.cs:8]: Exposed static helpers to construct the matcher based on what we want to match

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/NameMatcher.cs
Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/NameMatcher.cs:48]: Mostly because we only want to take in a type, but a types parent could be a type or namespace. Added a comment to that effect.

**Comment by twsouthwick** [comment]
The goal here is solely to replace usage of HttpContext.Current with method injection which is something that can be done before upgrading. Replacing HttpContext with an abstraction would be the next step, but I don't think we'd want to do that in this code fixer.  This enables the following scenario:

1. Find where HttpContext.Current is being used
2. Inject HttpContext rather than access via static method
3. Update calling patterns to pass through from controller/filter/etc
4. [TODO] Replace HttpContext with custom abstraction
5. Upgrade to ASP.NET Core
6. No HttpContext[.Current] usage in project so no need to deal with it now

The benefit of this is:

1. Gets the code base into a better place without static accessors
2. Allows for the code to be updated without breaking anything

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/CSharpHttpContextRefactorCodeFixProvider.cs
Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactor/CSharpHttpContextRefactorCodeFixProvider.cs:14]: Done. Using extension methods

**Review by mjrousos — APPROVED** [review]
This looks good to me. I think it might be hard to incorporate into UA (if we want to do that) since it is most useful if run before the TFM change and because the total number of diagnostics won't initially go down when the code fix is applied (since the user is expected to apply it iteratively, several times), but the analyzer and code fix provider look useful for manual use alongside UA.

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs:22]: 👍

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs:34]: Syntactically speaking, what is a member?  Let's update with a comment block.

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs:80]: This looks like it should use one of the helper methods identified above
`GetStringComparison`

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs:92]: nit: prefer name GetExistingParameter**By**Symbol

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs:85]: nit: get's the **first** instance of an existing parameter

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs:123]: nit: GetParentOperation**Where**

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SyntaxNodeExtensions.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SyntaxNodeExtensions.cs:23]: this looks nice, let's merge this PR first so I can reuse this in the PR for binaryFormatter

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/WorkspaceExtensions.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/WorkspaceExtensions.cs:28]: If I copy and paste the same class file from projectA into projectB then this will find the first instance right?

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs:19]: 👍

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs:49]: What's the advantage here to registering compilation?

Should we retrofit other analyzers with this pattern?

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs:72]: nit: try to put string comparisons later in the analyzer "bail out" flow. I don't have a doc recommending this but I'm basing that on string comparison always being slower than other conditionals we can run.

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactorCodeFixProvider.cs
Comment (KSchlobohm) [inline src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/HttpContextRefactorCodeFixProvider.cs:17]: 👍

**Review by KSchlobohm — APPROVED** [review]
added some nits, otherwise looks good

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs
Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs:80]: actually, this is an assembly name comparison which is case insensitive AFAIK (maybe depends on the file system, though)

Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/SymbolExtensions.cs:80]: Added a comment, though

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/WorkspaceExtensions.cs
Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers.Common/WorkspaceExtensions.cs:28]: No - it's still a separate document

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs
Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs:49]: Good question - I'm not sure which is better. This will make it so only the one we care about is actually registered. Not sure how much idfference that has

File: src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs
Comment (twsouthwick) [reply src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/HttpContextCurrentAnalyzer.cs:72]: The best comparison we have here is if Current is not the name, then we shouldn't continue at all. It's the best filter I can think of

**Pull request merged**
